pkg/engine/executor/executor_test.go

```diff
@@ -0,0 +1,134 @@
+package executor
+
+import (
+	"context"
+	"testing"
+
+	"github.com/grafana/loki/v3/pkg/engine/internal/types"
+	"github.com/grafana/loki/v3/pkg/engine/planner/physical"
+	"github.com/stretchr/testify/require"
+)
+
+func TestExecutor(t *testing.T) {
+	t.Run("pipeline fails if plan is nil", func(t *testing.T) {
+		pipeline := Run(context.TODO(), Config{}, nil)
+		err := pipeline.Read()
+		require.ErrorContains(t, err, "failed to execute pipeline: plan is nil")
+	})
+
+	t.Run("pipeline fails if plan has no root node", func(t *testing.T) {
+		pipeline := Run(context.TODO(), Config{}, &physical.Plan{})
+		err := pipeline.Read()
+		require.ErrorContains(t, err, "failed to execute pipeline: plan has no root node")
+	})
+}
+
+func TestExecutor_DataObjScan(t *testing.T) {
+	t.Run("is not implemented", func(t *testing.T) {
+		c := &Context{}
+		pipeline := c.executeDataObjScan(context.TODO(), &physical.DataObjScan{})
+		err := pipeline.Read()
+		require.ErrorContains(t, err, errNotImplemented.Error())
+	})
+}
+
+func TestExecutor_SortMerge(t *testing.T) {
+	t.Run("no inputs result in empty pipeline", func(t *testing.T) {
```

Comment:
nit: do we need these checks during execution, ideally we should never run into them if planning is already performing validation?